karspexet/ticket/views.py

```diff
@@ -11,14 +30,125 @@ def home(request):
         'upcoming_shows': upcoming_shows
     })
 
+
 def select_seats(request, show_id):
     show = Show.objects.get(id=show_id)
-    forms = [SeatingGroupFormSet(seating_group) for seating_group in show.venue.seatinggroup_set.all()]
+    reservation = _get_or_create_reservation_object(request, show)
+    taken_seats_qs = Reservation.active.exclude(pk=reservation.pk).filter(show=show)
+    _set_session_timeout(request)
+
+    if request.POST:
+        seat_params = _seat_specifications(request)
+        if _all_seats_available(taken_seats_qs, seat_params.keys()):
+            reservation.tickets = seat_params
+            reservation.save()
+            return redirect("payment", show_id=show_id)
+        else:
+            messages.error(request, "Some of your chosen seats are already taken")
+
+    taken_seats = set(map(int,set().union(*[r.tickets.keys() for r in taken_seats_qs.all()])))
+    forms = [
+        SeatingGroupFormSet(seating_group, taken_seats)
+        for seating_group
+        in show.venue.seatinggroup_set.all()
+    ]
 
     return render(request, "select_seats.html", {
         'show': show,
         'venue': show.venue,
         'forms': forms,
     })
 
-    # return render(request, 'enter_contact_details.html', {})
+def payment(request, show_id):
+    # TODO: verify that reservation_timeout has not passed
+    # TODO: call stripe and verify data
+    # * create account/customer object
+    # * create ticket objects
+    # * mark reservation as paid
+    # * send email to customer
+
+    if _session_expired(request):
+        messages.warning(request, "Your session has expired. Please start over from scratch.")
+        return redirect("select_seats", show_id=show_id)
+
+    _set_session_timeout(request)
+    show = Show.objects.get(pk=show_id)
+    reservation = _get_or_create_reservation_object(request, show)
+    seats = Seat.objects.filter(pk__in=reservation.tickets.keys())
+
+    return render(request, 'payment.html', {
+        'show': show,
+        'seats': seats,
+        'reservation': reservation,
+        'stripe_key': stripe_keys['publishable_key'],
+    })
+
+def process_payment(request, reservation_id):
```

Comment:
Please use [transaction.atomic](https://docs.djangoproject.com/en/1.11/topics/db/transactions/#django.db.transaction.atomic) here, either as a decorator or context manager (with statement). This should probably also be added to other places, but this seems like a sure fire thing to start with.

_(I know you can set transactions for all HTTP request, but even if you do, having it explicit right here is important since it's handling money.)_